src/nvim/os/pty_process_unix.c

```diff
@@ -115,6 +115,11 @@ int pty_process_spawn(PtyProcess *ptyproc)
   return status;
 }
 
+const char *pty_process_tty_name(PtyProcess *ptyproc)
+{
+  return ptsname(ptyproc->tty_fd);
```

Comment:
ubuntu, clang 5:

```
[215/284] Building C object src/nvim/CMakeFiles/nvim.dir/os/pty_process_unix.c.o
../src/nvim/os/pty_process_unix.c:120:10: warning: implicit declaration of function 'ptsname' is invalid in C99 [-Wimplicit-function-declaration]
  return ptsname(ptyproc->tty_fd);
         ^
../src/nvim/os/pty_process_unix.c:120:10: warning: this function declaration is not a prototype [-Wstrict-prototypes]
../src/nvim/os/pty_process_unix.c:120:10: warning: incompatible integer to pointer conversion returning 'int' from a function with result type 'const char *' [-Wint-conversion]
  return ptsname(ptyproc->tty_fd);
         ^~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
This patch sets `add_definitions(-D_GNU_SOURCE)` for clang:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4b021ad6e71b..b5b5b341dbd7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -290,13 +290,14 @@ option(LOG_LIST_ACTIONS "Add list actions logging" OFF)
-
 add_definitions(-DINCLUDE_GENERATED_DECLARATIONS)
-
-if(CMAKE_COMPILER_IS_GNUCXX AND CMAKE_SYSTEM_NAME STREQUAL "Linux")
+if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
   set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--no-undefined")
   set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--no-undefined")
   set(CMAKE_MODULE_LINKER_FLAGS "${CMAKE_MODULE_LINKER_FLAGS} -Wl,--no-undefined")
-
   # For O_CLOEXEC, O_DIRECTORY, and O_NOFOLLOW flags on older systems
   # (pre POSIX.1-2008: glibc 2.11 and earlier). #4042
+  # For ptsname() on clang. #6743
   add_definitions(-D_GNU_SOURCE)
 endif()
```

Reply:
Probably `_XOPEN_SOURCE` should be used instead. https://stackoverflow.com/a/5724485/152142

Reply:
Should be in `stdlib.h` (already included) according to https://www.gnu.org/software/libc/manual/html_node/Allocation.html#index-ptsname

Reply:
should it only be done for linux? `ptsname()` is posix, it should be available in more places.

Reply:
Not sure. All I know is that `man ptsname` has this:

```
SYNOPSIS
       #define _XOPEN_SOURCE       /* See feature_test_macros(7) */
       #include <stdlib.h>

       char *ptsname(int fd);

       #define _GNU_SOURCE         /* See feature_test_macros(7) */
       #include <stdlib.h>

       int ptsname_r(int fd, char *buf, size_t buflen);
```

And various posts on the web seem to pair `#define _XOPEN_SOURCE` with the include. Only explanation I found was the SO link in my comment above.